services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/structure/directory/decorator/CountryIndexingDecorator.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Corona-Warn-App
+ *
+ * SAP SE and all other contributors /
+ * copyright owners license this file to you under the Apache
+ * License, Version 2.0 (the "License"); you may not use this
+ *  file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package app.coronawarn.server.services.distribution.assembly.diagnosiskeys.structure.directory.decorator;
+
+import app.coronawarn.server.services.distribution.assembly.structure.Writable;
+import app.coronawarn.server.services.distribution.assembly.structure.WritableOnDisk;
+import app.coronawarn.server.services.distribution.assembly.structure.archive.Archive;
+import app.coronawarn.server.services.distribution.assembly.structure.directory.Directory;
+import app.coronawarn.server.services.distribution.assembly.structure.directory.DirectoryOnDisk;
+import app.coronawarn.server.services.distribution.assembly.structure.directory.IndexDirectory;
+import app.coronawarn.server.services.distribution.assembly.structure.directory.decorator.IndexDirectoryDecorator;
+import app.coronawarn.server.services.distribution.assembly.structure.file.FileOnDisk;
+import app.coronawarn.server.services.distribution.assembly.structure.util.ImmutableStack;
+import app.coronawarn.server.services.distribution.config.DistributionServiceConfig;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This decorator creates the central index file for a country, e.g. DE.
+ */
+public class CountryIndexingDecorator<T> extends IndexDirectoryDecorator<T, WritableOnDisk> {
+
+  /**
+   * Separate each entry in the index file with new line.
+   */
+  private static final String NEW_LINE_SEPARATOR = "\r\n";
+
+  /**
+   * the name of this index file.
+   */
+  private final String fileName;
+
+  /**
+   * Creates a new decorator instance for the given directory.
+   *
+   * @param directory                 The target country directory.
+   * @param distributionServiceConfig The config.
+   */
+  public CountryIndexingDecorator(IndexDirectory<T, WritableOnDisk> directory,
```

Comment:
Type of argument `directory` could be narrowed down to `DiagnosisKeysCountryDirectory` to avoid accidental misuse. The order of the decorators in `DiagnosisKeysDirectory.java` would need to be flipped then.